### DISCUSSION TITLE
The best way to program a pneumatic switch on a PUP L motor

### DISCUSSION THREAD

**Discussion by Flippoel** [pybricks/support#1998] [Q&A]
Hi, 
I'm totally new here at the forum and working with Pybricks, I hope you can help me here and there to get a nice program for our model of a [ship](https://www.bricklink.com/v3/studio/design.page?idModel=429645).
In this ship we made 3 pneumatic switches who will be operated bij PUP L motors.
What's the best way to program them?
We tried to do it by setting the angle, but it worked only into one way, it doesn't want to turn into -degrees.
We also tried with the steering function, but there with the calibration by starting it's not pushed to the and of what the switch needed, so there we using the switch only partially. We had this last problem also with the pup app, so there we pushed a few time on the red 'stop' button until it worked properly. One option is to do the syncing a few times, but I guess there is a better solution.

I will attache some files to show what we did.

Thanks for your help

<img width="1019" alt="Scherm­afbeelding 2024-12-30 om 14 11 40" src="https://github.com/user-attachments/assets/a3dd8307-f078-4e41-a6fe-dbab91b8482f" />
<img width="411" alt="Scherm­afbeelding 2024-12-30 om 14 21 38" src="https://github.com/user-attachments/assets/c0ebd1d3-d509-4af0-b6cf-9dfff4b24059" />
<img width="545" alt="Scherm­afbeelding 2024-12-30 om 14 35 07" src="https://github.com/user-attachments/assets/d5618b13-4ef5-4a94-81bb-7a5fed244015" />

![Spauwer 2 0](https://github.com/user-attachments/assets/f7e253a5-d206-4dac-88aa-f6989904cea6)

**Comment by BertLindeman** [comment]
Wow, what a model.

Great.

Do you mean a pneumatic switch like this on a PF motor?
![image](https://github.com/user-attachments/assets/ab789e43-e471-4138-a035-c7662aceae62)

You speak of setting the angle, but the PF-motors have only the duty-cycle to set the speed.

There are a few problems to overcome:
1. The PF motors have no rotation sensors, so we cannot determine if the motor stalls.
2. The pneumatic switches are stiff if not used for a while and a few movements help to get them moving more smooth.
3. [EDIT: add] To reach the left and right position is doable, but the mid position is tricky, due to no internal sensors.  

You could try to move the switch with a low duty-cycle `motor.dc(value)`
Valid dc values are 15, 30, 45, 60, 75, 90, 100. (Positive or negative)
Other values will be rounded down to the nearest match.

You probably already looked at the [stable PFMotor doc](https://code.pybricks.com/static/docs/v2.19.0/pupdevices/pfmotor.html)

Does this help you getting started?

Bert

**Reply by Flippoel** [reply]
Hi Bert,

Thanks for your reply. No, we are not using PF motors. We use powered up motors as shown in the picture.
We are aware of the 'stiffness' of the switches, so moving it a few times at start of the program can be helpful. But how to program that?
<img width="1019" alt="Scherm­afbeelding 2024-12-30 om 14 11 40" src="https://github.com/user-attachments/assets/a880c862-fc20-4939-b938-e86fff0d9b2a" />
![IMG_3759](https://github.com/user-attachments/assets/0801606c-030f-41b9-97d7-3a5f619ede60)

**Reply by laurensvalk** [reply]
Thanks for joining the forum!

Are the motors still accessible in your build?

Despite having no markers, these motors do have a an absolute 0 point. You could simplify your code a lot if you set them to 0, and then build it in.

To run a motor to zero, just use the run to target angle block, and choose 0.

Then when built in, you can simply go to one of the 3 positions, for example 0, - 60, +60.

**Reply by BertLindeman** [reply]
Thanks Laurens, 

I did not know that motor also has an abs 0 point.

I tried to get the motor to zero, but the resistance of the switch is high.

**Comment by BertLindeman** [comment]
Oops, sorry.

That is better, those motors **do** have internal sensors.

Thinking out loud (no programming language, just thinking):
setup / preparation:
1. on start wiggle the switch a few times (try how many times) 
2. run the motor to one side until it stalls to determine one endpoint
3. run the motor the other way until it stalls to get the other endpoint
4. The middle of the two is the mid position of the switch
5. remember left / right and middle positions


To get started with detecting the switch endpoints, look at [doc about run_until_stalled](https://code.pybricks.com/static/docs/v2.19.0/pupdevices/motor.html#pybricks.pupdevices.Motor.run_until_stalled)

In that same doc there are [example Stall examples](https://code.pybricks.com/static/docs/v2.19.0/pupdevices/motor.html#stall-examples)

Let me know your progress.

Bert

**Comment by Flippoel** [comment]
Hi @BertLindeman and @laurensvalk,

Thanks a lot for your help, I tried this setup and program and that seems to work. 

<img width="719" alt="Scherm­afbeelding 2024-12-31 om 16 26 01" src="https://github.com/user-attachments/assets/667bcae1-32ca-4c12-a5f5-ced1482923c1" />

**Reply by BertLindeman** [reply]
Hi (Flip, I think)

Just ran this here (without the great ship, pitty) and it runs fine!
Well done.

New challenge: 😃  You could now make your own task(s) so you do not need to copy this motor on port.A for the other motors.

Bert
